Add recipe solo-jazz-theme

cstby opened the pull request:
### Brief summary of what the package does

An original theme inspired by the 1992 Solo Jazz cup design.

### Direct link to the package repository

https://github.com/cstby/solo-jazz-emacs-theme

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

None Needed.

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings **See notes below**
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### Notes
- M-x checkdoc advsied me to disambiguate `rainbow-mode`, but I saw no way to do so that passed the check.

riscy commented:
Thanks, I like this.  Here's a quick first pass.

### solo-jazz-theme.el ###

byte-compile (using Emacs 27.1):
```
solo-jazz-theme.el:94:1:Error: Symbol's value as variable is void: solo-jazz-default-colors-alist
```

checkdoc (using version 0.6.2):
- No issues!

check-declare-file (optional):
- No issues!

package-lint (using version 20200906.512):
```
3 issues found:
1:0: warning: Including "Emacs" in the package summary is usually redundant.
1:0: warning: The package summary should not end with a period.
1:0: error: There is no (provide 'solo-jazz-theme) form.
```
The latter (no `provide`) is technically redundant, but it's easy enough to pacify package-lint.

Your `declare-function`s will pacify the byte compiler, but they're actually malformed -- the second argument should be a string indicating a filename.  This is a non-blocker.

Other possible lints:
- solo-jazz-theme.el#L575: It's safer to sharp-quote function names; use `#'`
- solo-jazz-theme.el#L21: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))

riscy commented:
Feel free to ping when ready for another look!

cstby commented:
Thanks for the ping!

I just pushed my new changes to my master branch. I believe I've adressed everything you mentioned above, but I'm still a little puzzled by the byte-compile error. I believe it is fixed; I'm no longer getting this error when I byte-compile the file. However, flycheck is still giving me that same error. I haven't yet determined why this is the case. Any advice would be appreicated.

riscy commented:
I think the byte-compile is a [wrong time](https://www.gnu.org/software/emacs/manual/html_node/elisp/Wrong-Time.html#Wrong-Time) issue.  I suspect your Emacs instance has already loaded the code and in particular that variable, while flycheck (helpfully!) does a clean-room compile.

I also notice package-lint caught the `(provide 'solo-jazz-theme)` but now it knows that your package is called `solo-jazz-theme` and wants the prefixes to align -- should be an easy fix.

This is a bit risky:
```
(when (fboundp 'advice-add)
  (advice-add 'rainbow-turn-on :after  #'solo-jazz--rainbow-turn-on)
  (advice-add 'rainbow-turn-off :after #'solo-jazz--rainbow-turn-off))
```
`advice-add` should always be bound, so this will always run.  More problematically, if someone switches to another theme, this advice will continue to lurk behind the scenes.  If the user runs `unload-feature` on your package, errors will be thrown when they try to `rainbow-turn-on/off`.  What I'd recommend is leaving the functions in place and documenting the advice-adding as something your users can do if they want.  If you want to make it easier for them, you could support it with a minor mode.

cstby commented:
I've changed the prefixes to align with the new package name. I've also removed the risky `advice-add`s and added documentation in the readme about how to achieve that effect. I imagine that most users do not need it.

I fixed the byte-compile error by wrapping the color alists in `eval-and-complie`. Let me know if you think that's an okay approach? As a result, I'm seeing warnings about unused variables because the colors alist is used in two places, the second of which does not use all of the bound colors.

riscy commented:
To solve the byte-compile warnings, could you just put both blocks within the same macro-expansion of `solo-jazz-theme-with-color-variables`?

cstby commented:
Sounds good. I'll want to test whether that approach would allow for color overrides. I'm okay with removing that feature if needed. I'll reply here after I've pushed that change (and any necessary documentation).

cstby commented:
I apologize about the long delay. I got married recently and was out on honeymoon. I'm back now and should be able to respond promptly.

I've put both block in the same macroexpansion as you suggested. I've also removed the override functionality (including documentation).

riscy commented:
Hi @cstby, it's my turn to apologize.  Let me remove the awaiting upstream label and I'll take a look soon.

riscy commented:
Welcome to MELPA. :)
